src/plugins/discover/public/application/main/utils/fetch_all.ts

```diff
@@ -165,6 +167,54 @@ export function fetchAll(
   }
 }
 
+export async function fetchMoreDocuments(
+  dataSubjects: SavedSearchData,
+  fetchDeps: FetchDeps
+): Promise<void> {
+  try {
+    const { getAppState, getInternalState, services, savedSearch } = fetchDeps;
+    const searchSource = savedSearch.searchSource.createChild();
+
+    const dataView = searchSource.getField('index')!;
+    const query = getAppState().query;
+    const recordRawType = getRawRecordType(query);
+
+    if (recordRawType === RecordRawType.PLAIN) {
+      // not supported yet
+      return;
+    }
+
+    const lastDocuments = dataSubjects.documents$.getValue().result || [];
+    const lastDocumentSort = lastDocuments[lastDocuments.length - 1]?.raw?.sort;
+
+    if (!lastDocumentSort) {
+      return;
+    }
+
+    searchSource.setField('searchAfter', lastDocumentSort);
+
+    // Mark as loading
+    sendLoadingMoreMsg(dataSubjects.documents$);
+
+    // Update the searchSource
+    updateVolatileSearchSource(searchSource, {
+      dataView,
+      services,
+      sort: getAppState().sort as SortOrder[],
+      customFilters: getInternalState().customFilters,
+    });
+
+    // Fetch more documents
+    const { records } = await fetchDocuments(searchSource, fetchDeps);
+
+    // Update the state and finish the loading state
+    sendLoadingMoreFinishedMsg(dataSubjects.documents$, records);
+  } catch (error) {
+    sendLoadingMoreFinishedMsg(dataSubjects.documents$, []);
+    sendErrorMsg(dataSubjects.main$, error);
+  }
+}
```

Comment:
@davismcphee 

> Why the double try/catch logic? Would it work to instead make the function async and have a single try/catch handle the whole function?

> Instead of showing the errors in toasts, we could call sendErrorMsg(dataSubjects.main$, error); to show them in the inline callout.

Updated via https://github.com/elastic/kibana/pull/163784/commits/b31b568dfd9a5d0dba0eee8c226bd7ab62fd73b2, thanks for the idea!

Reply:
Also added inlining for the shard warnings via https://github.com/elastic/kibana/pull/163784/commits/3fdf229fe1371a95bc50b63e7c156d148441590a